### ISSUE TITLE
Determine when child process terminates unexpectedly (crashes)

### ISSUE THREAD

**Issue by johnib** [Tyrrrz/CliWrap#85]
Hello,

I'm using CliWrap with the following flavor:

```CSharp
            Command command = Cli.Wrap(filePath)
                .WithArguments(arguments)
                .WithValidation(CommandResultValidation.ZeroExitCode);

            _logger.LogInformation("Spawning process with command: {command}", command);
            await command.Observe(Encoding.UTF8, stoppingToken)
                .ForEachAsync(commandEvent =>
```

Is there any way I can determine the child process being spawned by CliWrap crashes? (which in that case I'd want to restart it..)
Initially I expected to it to be covered by the following case:

```CSharp
                        case ExitedCommandEvent exitedCommandEvent:
                            _logger.LogCritical("Process exited with code: {exitCode}", exitedCommandEvent.ExitCode);
                            break;
```

However, this only catches the case of graceful termination.

Hope my question is clear..

Thanks for helping!

**Comment by Tyrrrz** [comment]
Hi.

Hm. The exited event should definitely cover non-graceful termination as well. What do you experience instead? Does it hang?

**Comment by johnib** [comment]
Hi

I've put a breakpoint on the 'exit' event, but it was never called and infect the main process just hanged, waiting for more events from the Observable. 

The way the process terminated was by me killing it via Windows' Task Manager.

**Comment by Tyrrrz** [comment]
Can you see if it also hangs when you use other execution models (i.e. `ExecuteAsync()`/`ListenAsync()`)?

**Comment by johnib** [comment]
With `ExecuteAsync()` I get the same "hanging" behavior:

```CSharp
            Command command = Cli.Wrap(filePath)
                .WithArguments(arguments)
                .WithValidation(CommandResultValidation.ZeroExitCode);

            var result = await command.ExecuteAsync(stoppingToken);
            _logger.LogInformation("Result: {result}", result);
```

**Comment by Tyrrrz** [comment]
I just tried the following code:

```csharp

        [Fact]
        public async Task Test()
        {
            // Arrange
            var cmd = Cli.Wrap("dotnet")
                .WithArguments(a => a
                    .Add(Dummy.Program.FilePath)
                    .Add(Dummy.Program.Sleep)
                    .Add(100_000));

            // Act
            await cmd.ExecuteAsync();

            // Assert
            Assert.True(true);
        }
```

This is using the dummy program from tests and this command causes it to sleep for 100 seconds. While it was waiting, I killed it with task manager and `await cmd.ExecuteAsync()` completed with an exception (exit code not set to zero).

It seems I can't reproduce your issue. Could it be that there's something special in your setup?

**Comment by johnib** [comment]
This is weird.. are you running on Linux? or Windows?

Can you try setting up a simple console app, spawning some process, and kill it as well via Task Manager?
(just don't run it as a unit test via mstest/whatever)

**Comment by Tyrrrz** [comment]
Yes, I'm running it on Windows. Console or test shouldn't make a difference, but I'll try.

**Comment by Tyrrrz** [comment]
Tried it on a console, works the same way.

Can you try the same thing but with bare `System.Diagnostics.Process`? That's what CliWrap uses under the hood and relies on it to report when the process is terminated. [Link](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.process.exited?view=netcore-3.1#remarks)

![image](https://user-images.githubusercontent.com/1935960/93455331-e5366280-f8e4-11ea-8ccf-e6e0b336f0fd.png)

**Comment by vivainio** [comment]
I am having this same problem with one particular process. It's launched with "cmd /c mycommand.cmd". It stops yielding events when it should stop. It's a normal cmd file that exits with zero status.

My code:

```csharp
            await foreach (var cmdEvent in cmd.ListenAsync())
            {
                switch (cmdEvent)
                {
                    case StartedCommandEvent start:
                        write(v, cname, "Start pid " + start.ProcessId);
                        RunningPid[index] = start.ProcessId;
                        break;
                    case StandardOutputCommandEvent stdOut:
                        write(v, cname, stdOut.Text); 
                        break;
                    case StandardErrorCommandEvent stdErr:
                        write(v, "red", stdErr.Text); 
                        break;
                    case ExitedCommandEvent ex:
                        return (ex.ExitCode, 0);
                    default:
                        write("UNK", "red", "Unknown event :" + cmdEvent.ToString());
                        break;
                }
            }
```

**Comment by Tyrrrz** [comment]
> @Tyrrrz do you think this is enough information to reopen the ticket? I can do further diagnostics if needed

Yes, please do.

**Comment by vivainio** [comment]
> > @Tyrrrz do you think this is enough information to reopen the ticket? I can do further diagnostics if needed
> 
> Yes, please do.

I think you need to do it, I don't have the rights

**Comment by vivainio** [comment]
@Tyrrrz I created a new ticket instead of bothering with this closed one, #185

**Comment by Tyrrrz** [comment]
> @Tyrrrz I created a new ticket instead of bothering with this closed one, #185

Yes, that's what I meant. Thank you.
